File: app/src/main/java/life/plank/juna/zone/data/local/db/RoomDb.java
package life.plank.juna.zone.data.local.db;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;

import life.plank.juna.zone.ZoneApplication;
import life.plank.juna.zone.data.model.FootballTeam;

@Database(entities = {FootballTeam.class}, version = 1, exportSchema = false)
public abstract class RoomDb extends RoomDatabase {

    private static final String ROOM_DB_NAME = "junaDatabase.db";
    private static RoomDb INSTANCE;

    RoomDb() {
//        Reflection safety
        if (INSTANCE != null) {
            throw new RuntimeException("Use getInstance() method to get the single instance of this class");
        }
    }

    public static RoomDb getInstance() {
        if (INSTANCE == null) {
//            synchronized block and a second null-check for thread safety
            synchronized (RoomDb.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(ZoneApplication.getContext(), RoomDb.class, ROOM_DB_NAME).build();
                }
            }
        }
        return INSTANCE;
    }

    //    returning same instance when serializing/de-serializing
    protected RoomDb readResolve() {
        return getInstance();
    }
}

File: app/src/main/java/life/plank/juna/zone/data/local/db/RoomDb.kt
package life.plank.juna.zone.data.local.db

import androidx.room.*
import life.plank.juna.zone.ZoneApplication
import life.plank.juna.zone.data.local.dao.*
import life.plank.juna.zone.data.model.football.LeagueInfo
import life.plank.juna.zone.data.local.typeconverter.*
import life.plank.juna.zone.data.model.football.MatchDetails

@Database(version = 1, exportSchema = false, entities = [
    MatchDetails::class,
    LeagueInfo::class
])
@TypeConverters(
        DateConverter::class,
        LineupsConverter::class,
        MatchConverter::class
)
abstract class RoomDb : RoomDatabase() {

    abstract fun matchDetailsDao(): MatchDetailsDao
    abstract fun leagueDao(): LeagueDao

    companion object {
        private const val ROOM_DB_NAME = "junaDatabase.db"
        private var INSTANCE: RoomDb? = null

        //        synchronized block and a second null-check for thread safety
        fun getInstance(): RoomDb {
            if (INSTANCE == null) {
                synchronized(RoomDb::class) {
                    if (INSTANCE == null) {
                        INSTANCE = Room.databaseBuilder(ZoneApplication.getContext(), RoomDb::class.java, ROOM_DB_NAME).build()
                    }
                }
            }
            return INSTANCE!!
        }

        fun destroyInstance() {
            INSTANCE = null
        }
    }
}